quickjs.c

```diff
@@ -51374,12 +51540,95 @@ static JSValue js_array_buffer_transfer(JSContext *ctx,
     abuf->data = NULL;
     abuf->byte_length = 0;
     abuf->detached = TRUE;
-    return js_array_buffer_constructor3(ctx, JS_UNDEFINED, new_len,
+    return js_array_buffer_constructor3(ctx, JS_UNDEFINED, new_len, pmax_len,
                                         JS_CLASS_ARRAY_BUFFER,
                                         bs, abuf->free_func,
                                         NULL, FALSE);
 }
 
+static JSValue js_array_buffer_resize(JSContext *ctx, JSValue this_val,
+                                      int argc, JSValue *argv, int class_id)
+{
+    uint32_t size_log2, size_elem;
+    struct list_head *el;
+    JSArrayBuffer *abuf;
+    JSTypedArray *ta;
+    JSObject *p;
+    uint8_t *data;
+    int64_t len;
+
+    abuf = JS_GetOpaque2(ctx, this_val, class_id);
+    if (!abuf)
+        return JS_EXCEPTION;
+    if (abuf->shared) {
+        if (class_id == JS_CLASS_ARRAY_BUFFER)
+            return JS_ThrowTypeError(ctx, "resize called on SharedArrayBuffer");
+    } else {
+        if (class_id == JS_CLASS_SHARED_ARRAY_BUFFER)
+            return JS_ThrowTypeError(ctx, "grow called on ArrayBuffer");
+    }
+    if (abuf->detached)
+        return JS_ThrowTypeErrorDetachedArrayBuffer(ctx);
+    if (!array_buffer_is_resizable(abuf))
+        return JS_ThrowTypeError(ctx, "array buffer is not resizable");
+    // TODO(bnoordhuis) support externally managed RABs
+    if (abuf->free_func != js_array_buffer_free)
+        return JS_ThrowTypeError(ctx, "external array buffer is not resizable");
+    if (JS_ToInt64(ctx, &len, argv[0]))
+        return JS_EXCEPTION;
+    if (len < 0 || len > abuf->max_byte_length) {
+    bad_length:
+        return JS_ThrowRangeError(ctx, "invalid array buffer length");
+    }
+    // SABs can only grow and we don't need to realloc because
+    // js_array_buffer_constructor3 commits all memory upfront;
+    // regular RABs are resizable both ways and realloc
+    if (abuf->shared) {
+        if (len < abuf->byte_length)
+            goto bad_length;
+        // Note this is off-spec; there's supposed to be a single atomic
```

Comment:
Out of curiosity: how is this supposed to work? As in, how would t1 let t0 know that they grew the SAB? Some kind of shared memory region where the length is stored or?